sql: support DEFAULT expressions for routine parameters

yuzefovich opened the pull request:
This commit adds support for DEFAULT expressions for routine parameters.
Routines can be declared with default values for some or all input
arguments. The default values are inserted whenever the routine is
called with insufficiently many actual arguments. Since arguments can
only be omitted from the end of the actual argument list, all
parameters after a parameter with a default value have to have default
values as well. For procedures additionally no OUT parameters are
allowed after the first input parameter with the DEFAULT expression. The
DEFAULT expression has to be coercible to the argument type of the
parameter.

Given that presence of DEFAULT expressions changes the overload
resolution, we need to include these into `FunctionSignature` proto. The
overload resolution logic has also been adjusted to try omitting some of
the input types for comparison (when applicable). In particular, UDF
dependency tracking for memo metadata has been extended to store the
"invocation signature" in addition to the overload itself given that
input types of the latter might now be different from the former.

In order to correctly track dependencies during schema changes, we
always serialize the DEFAULT expressions after having them type-checked.

Fixes: #100962.

Release note (sql change): DEFAULT expressions for input parameters of
user-defined functions and stored procedures are now supported.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121082)
<!-- Reviewable:end -->

yuzefovich commented:
There are a few unresolved questions that I'm not sure about and I need some help with. In particular,
- Drew, Marcus, in the optbuilder I have two TODOs of the form `TODO(check with reviewers)`
- Faizan, I have a handful of `TODO(100962)` in the new schemachanger. In particular, I'm not quite sure how I'm supposed to use `FunctionParamDefaultExpression`.

Other than that, it's RFAL.

DrewKimball reviewed (commented):
Nice work! Very thorough.

> Reviewed 46 of 46 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-:-Nu93-ZA0nc0uW-TFK-J:b-nfpvzg)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @mgartner, and @yuzefovich)

___
*[`pkg/sql/opt/metadata.go` line 1181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8mIHY-DM8-YXnN6jF:-Nu8mIHY-DM8-YXnN6jG:b-67veke) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/metadata.go#L1181)):*
> ```Go
> 			return false
> 		}
> 	}
> ```

[nit] We should also check that the lengths of the maps are equal.
___
*[`pkg/sql/opt/optbuilder/create_function.go` line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8o36o-iKldNh-CnPf:-Nu8o36o-iKldNh-CnPg:bxeums) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/optbuilder/create_function.go#L234)):*
> ```Go
> 				if !cast.ValidCast(resolved, typ, cast.ContextAssignment) {
> 					panic(pgerror.Newf(pgcode.DatatypeMismatch,
> 						"unable to coerce type %s to %s", resolved.Name(), typ.Name(),
> ```

This looks like the right way to handle coercion.

[nit] I think postgres uses the message `argument of DEFAULT must be type %s, not type %s`
relevant code link: https://github.com/postgres/postgres/blob/0075d78947e3800c5a807f48fd901f16db91101b/src/backend/parser/parse_coerce.c#L1223-L1233
___
*[`pkg/sql/opt/optbuilder/create_function.go` line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8pXwN-nqzh0WWWsZC:-Nu8pXwN-nqzh0WWWsZD:bf44v8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/optbuilder/create_function.go#L242)):*
> ```Go
> 			// TODO(check with reviewers): is this the right way to do that?
> 			// We also build the expression later, during execution of the
> 			// routine - should we store this scalar expression somewhere?
> ```

This seems right to me as well - we do the same thing for the body statements.
___
*[`pkg/sql/opt/optbuilder/routine.go` line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8rdXJ7X3GbQaR7fr-:-Nu8rdXK3xewM29Co4ne:bi1stnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/optbuilder/routine.go#L303)):*
> ```Go
> 					panic(err)
> 				}
> 				pexpr := inScope.resolveType(param.DefaultVal, typ)
> ```

It's probably not possible to reach this point, but I think using `inScope` here would allow you to write a default expression like `a + 5`, and if the routine was called in a context where some column `a` was in scope, it would resolve as that column. That's probably not what we want, so we should just use `tree.TypeCheck` directly.
___
*[`pkg/sql/sem/tree/overload.go` line 395 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8zcxNBXU6VXuI1QxB:-Nu8zcxNBXU6VXuI1QxC:b-x9chy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/sem/tree/overload.go#L395)):*
> ```Go
> 	// Overload.OutParamOrdinals, Overload.OutParamTypes, and
> 	// Overload.DefaultExprs for more details.
> 	extraParamInfo() (_ RoutineType, outParamOrdinals []int32, outParamTypes TypeList, defaultExprs Exprs)
> ```

[nit] maybe at some point we should just split this into separate methods
___
*[`pkg/ccl/logictestccl/testdata/logic_test/procedure_params` line 560 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu920Qd8YRGKL4qRDum:-Nu920Qd8YRGKL4qRDun:b-kvdln9) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/ccl/logictestccl/testdata/logic_test/procedure_params#L560)):*
> ```
> param
> (1,2)
> ```

It might be interesting to also test a narrowing type coercion:
```
postgres=# CREATE FUNCTION f(x CHAR DEFAULT 'foo') RETURNS CHAR LANGUAGE SQL AS $$ SELECT x; $$;
CREATE FUNCTION
postgres=# SELECT f();
  f
-----
 foo
(1 row)
```
Note that postgres doesn't actually truncate the value, but I think we will. I think that's fine, and we track that behavior for PL/pgSQL coercion here: https://github.com/cockroachdb/cockroach/issues/115385
___
*[`pkg/ccl/logictestccl/testdata/logic_test/procedure_params` line 561 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu92RIj9hAL-m4soAgi:-Nu92RIj9hAL-m4soAgj:b37pjgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/ccl/logictestccl/testdata/logic_test/procedure_params#L561)):*
> ```
> (1,2)
> 
> subtest default_exprs
> ```

Can we do a test like:
```
CREATE FUNCTION f(x INT, y INT DEFAULT x + 1) ...
```


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/logictest/testdata/logic_test/udf_params:
```diff
@@ -593,3 +593,147 @@ SELECT b FROM f_udt()
+CREATE FUNCTION f1(a INT, b INT = 2) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b; $$;
+
+statement ok
+CREATE FUNCTION f2(a INT, b INT = f1(1)) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b; $$;
```
can you also add tests for having a default for an enum type and composite type?

that should also be tested in the random schema change workload; not sure if you want to do that now, or file a ticket for later.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go` line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-NuPuIi7E5eSipWkgupE:-NuPuIi7E5eSipWkgupF:bbepqhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go#L117)):*
> ```Go
> 		if param.DefaultVal != nil {
> 			fn.Params[i].DefaultExpr = tree.Serialize(param.DefaultVal)
> 			// TODO(100962): probably need a similar call to b.WrapFunctionBody
> ```

We probably want something closer to: WrapExpression, I'm guessing parameter references are allowed in the expression. I think if we do that and register the attributes on attr.go things should sort of work with minimal changes for dependency tracking


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: support DEFAULT expressions for routine parameters

This commit adds support for DEFAULT expressions for routine parameters.
Routines can be declared with default values for some or all input
arguments. The default values are inserted whenever the routine is
called with insufficiently many actual arguments. Since arguments can
only be omitted from the end of the actual argument list, all
parameters after a parameter with a default value have to have default
values as well. For procedures additionally no OUT parameters are
allowed after the first input parameter with the DEFAULT expression. The
DEFAULT expression has to be coercible to the argument type of the
parameter.

Given that presence of DEFAULT expressions changes the overload
resolution, we need to include these into `FunctionSignature` proto. The
overload resolution logic has also been adjusted to try omitting some of
the input types for comparison (when applicable). In particular, UDF
dependency tracking for memo metadata has been extended to store the
"invocation signature" in addition to the overload itself given that
input types of the latter might now be different from the former.

In order to correctly track dependencies during schema changes, we
always serialize the DEFAULT expressions after having them type-checked.

Release note (sql change): DEFAULT expressions for input parameters of
user-defined functions and stored procedures are now supported.

yuzefovich reviewed (commented):
Thanks everyone for the reviews! I think there is only one unresolved question in my mind about the new schemachanger - why do we have this `FunctionParamDefaultExpression`? Can we get rid off it? To me it seems like this element is redundant.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-:-NuWK6Nd313ibrNe7daN:bss3559)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @fqazi, @mgartner, and @rafiss)

___
*[`pkg/sql/opt/metadata.go` line 1181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8mIHY-DM8-YXnN6jF:-NuR9-3t-AJyuaaa8QUM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/metadata.go#L1181)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] We should also check that the lengths of the maps are equal.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/create_function.go` line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8o36o-iKldNh-CnPf:-NuRigllDenLBHsUebl9:b-9p766c) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/optbuilder/create_function.go#L234)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This looks like the right way to handle coercion.

[nit] I think postgres uses the message `argument of DEFAULT must be type %s, not type %s`
relevant code link: https://github.com/postgres/postgres/blob/0075d78947e3800c5a807f48fd901f16db91101b/src/backend/parser/parse_coerce.c#L1223-L1233
</blockquote></details>

Thanks. I changed the error message but I'm struggling to actually come up with a reproduction for it (I'm always getting `invalid input syntax for type` errors instead).
___
*[`pkg/sql/opt/optbuilder/create_function.go` line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8pXwN-nqzh0WWWsZC:-NuRggNw9v4v6P0h76kH:be17u8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/optbuilder/create_function.go#L242)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This seems right to me as well - we do the same thing for the body statements.
</blockquote></details>

Ack, thanks.
___
*[`pkg/sql/opt/optbuilder/routine.go` line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8rdXJ7X3GbQaR7fr-:-NuWEOmH5Qk4-l-MfGxx:bo3vl6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/optbuilder/routine.go#L303)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It's probably not possible to reach this point, but I think using `inScope` here would allow you to write a default expression like `a + 5`, and if the routine was called in a context where some column `a` was in scope, it would resolve as that column. That's probably not what we want, so we should just use `tree.TypeCheck` directly.
</blockquote></details>

Thanks, done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go` line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-NuPuIi7E5eSipWkgupE:-NuRaM8847PLzZL6ppmx:b-xhskgf) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go#L117)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

We probably want something closer to: WrapExpression, I'm guessing parameter references are allowed in the expression. I think if we do that and register the attributes on attr.go things should sort of work with minimal changes for dependency tracking
</blockquote></details>

Thanks for the pointer - this looks like exactly what we need. DEFAULT expressions can reference sequences and UDFs, but cannot reference columns or other parameters.

I still don't quite understand what the benefit of having `FunctionParamDefaultExpression` in the first place - why do we need it? It seems to me if we just remove it, everything will keep on working as is.
___
*[`pkg/sql/sem/tree/overload.go` line 395 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8zcxNBXU6VXuI1QxB:-NuRBB9QApZRJkgo68gH:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/sem/tree/overload.go#L395)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] maybe at some point we should just split this into separate methods
</blockquote></details>

Good point, done.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/procedure_params` line 560 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu920Qd8YRGKL4qRDum:-NuRIa9T52i5JPuTQlMU:bld8h7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/ccl/logictestccl/testdata/logic_test/procedure_params#L560)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It might be interesting to also test a narrowing type coercion:
```
postgres=# CREATE FUNCTION f(x CHAR DEFAULT 'foo') RETURNS CHAR LANGUAGE SQL AS $$ SELECT x; $$;
CREATE FUNCTION
postgres=# SELECT f();
  f
-----
 foo
(1 row)
```
Note that postgres doesn't actually truncate the value, but I think we will. I think that's fine, and we track that behavior for PL/pgSQL coercion here: https://github.com/cockroachdb/cockroach/issues/115385
</blockquote></details>

Done. This actually exposed a bug in `CheckDependencies` that is now fixed.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/procedure_params` line 561 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu92RIj9hAL-m4soAgi:-NuRDSf2-eLv9HpiZWL5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/ccl/logictestccl/testdata/logic_test/procedure_params#L561)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can we do a test like:
```
CREATE FUNCTION f(x INT, y INT DEFAULT x + 1) ...
```
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_params` line 715 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#gh-1544611464:-NuWI38r9tBC3VUakg-k:bnfqb43) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/logictest/testdata/logic_test/udf_params#L715)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you also add tests for having a default for an enum type and composite type? (and more importantly, make sure neither the type _nor_ the specific value from the enum can be dropped while the function default param is still using it.)

that should also be tested in the random schema change workload; not sure if you want to do that now, or file a ticket for later.
</blockquote></details>

Thanks, added tests for this which uncovered a few bugs. I'll look into the schema changer workload to see if it's easy to add this extension there.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
schemachanger: remove FunctionParamDefaultExpression as separate element

There is no advantage to tracking the DEFAULT expression of routine
parameters as a separate element since this expression can only be
changed when modifying the whole function (and cannot be dropped once
added). Instead, we'll simply store the expressions in the serialized
form within the function parameters.

Release note: None

yuzefovich commented:
I opened #121587 to see what breaks if I simply remove `scpb.FunctionParamDefaultExpression`.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, @rafiss, and @yuzefovich)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go` line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-NuPuIi7E5eSipWkgupE:-NuWYI7FAAoVQuTIhuYJ:b-lnp8if) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go#L117)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Thanks for the pointer - this looks like exactly what we need. DEFAULT expressions can reference sequences and UDFs, but cannot reference columns or other parameters.

I still don't quite understand what the benefit of having `FunctionParamDefaultExpression` in the first place - why do we need it? It seems to me if we just remove it, everything will keep on working as is.
</blockquote></details>

I think the only advantage is if we want to have some alter without dropping the entire function.  We normally just replace the entire function, so I think we can drop it as long as the dependencies are tracked with the parameters


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
Queries logic <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 23 of 24 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-:-NuWvqPv6UcUCLWySVQI:b-72r6mr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @yuzefovich)

___
*[`pkg/sql/opt/optbuilder/create_function.go` line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8o36o-iKldNh-CnPf:-NuWuKJU2vq49YMiilSS:b-5snarq) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/optbuilder/create_function.go#L234)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Thanks. I changed the error message but I'm struggling to actually come up with a reproduction for it (I'm always getting `invalid input syntax for type` errors instead).
</blockquote></details>

This should do it:
```
create procedure p(x int default true) language sql as $$ select 1 $$;
```
or:
```
create procedure p(x int default ARRAY[1, 2]) language sql as $$ select 1 $$;
```
You have to make sure the assignment-cast isn't possible.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @fqazi, @mgartner, and @rafiss)

___
*[`pkg/sql/opt/optbuilder/create_function.go` line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nu8o36o-iKldNh-CnPf:-Nu_5U3k8uxqAQYWO1Np:b-pc6ok) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/opt/optbuilder/create_function.go#L234)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This should do it:
```
create procedure p(x int default true) language sql as $$ select 1 $$;
```
or:
```
create procedure p(x int default ARRAY[1, 2]) language sql as $$ select 1 $$;
```
You have to make sure the assignment-cast isn't possible.
</blockquote></details>

Thanks, done (for some reason, I only tried different numbers and strings, between which we do have assignment casts). Adding the test case actually exposed a bug.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go` line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-NuPuIi7E5eSipWkgupE:-Nu_5ed_Bg9MxI3g0kcb:b-3i1hra) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go#L117)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think the only advantage is if we want to have some alter without dropping the entire function.  We normally just replace the entire function, so I think we can drop it as long as the dependencies are tracked with the parameters
</blockquote></details>

Thanks, this makes sense - I added another commit that removes `FunctionParamDefaultExpression`, PTAL.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_params` line 715 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121082#gh-1544611464:-Nu_60NaA1krJOHEzHck:b-1kdpmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/be4f15fbb75dc7484835fd959e8d34e34c3a43c1/pkg/sql/logictest/testdata/logic_test/udf_params#L715)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Thanks, added tests for this which uncovered a few bugs. I'll look into the schema changer workload to see if it's easy to add this extension there.
</blockquote></details>

I briefly looked into a recent [change](https://github.com/cockroachdb/cockroach/commit/826ee6efdd55f36658641aa7355ff7395fa2d044) for supporting UDFs invoking other UDFs, and it doesn't seem exactly trivial, so I filed #121718.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@fqazi @rafiss could you give this PR another look (mostly at the first commit)?

fqazi reviewed (changes requested):
> Reviewed 3 of 46 files at r1, 61 of 61 files at r4, 2 of 45 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @rafiss, and @yuzefovich)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 681 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-NuecjJr3fwexDQtyLj3:-NuecjJr3fwexDQtyLj4:b-ks79zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/220f1a84eef868ffa23427ced83aea50ac917cd7/pkg/sql/schemachanger/scbuild/builder_state.go#L681)):*
> ```Go
> 			// disallowed.
> 			_, _, parentNamespace := scpb.FindNamespace(b.QueryByID(tableID))
> 			if parentNamespace != nil && desc.GetParentID() != parentNamespace.DatabaseID {
> ```

When did we run into this? This shouldn't be a thing :grimacing: 
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 672 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nued7vuCQt5U-S6a2bD:-Nued7vuCQt5U-S6a2bE:bpfajo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/220f1a84eef868ffa23427ced83aea50ac917cd7/pkg/sql/schemachanger/scpb/elements.proto#L672)):*
> ```proto
>     cockroach.sql.catalog.catpb.FunctionParamClass class = 2 [(gogoproto.nullable) = false];
>     TypeT type = 3 [(gogoproto.nullable) = false];
>     string default_expr = 4;
> ```

This won't work correctly, we can't track dependencies, we need to at least embed an expression directly in parameter. Otherwise, we can have references in the default expression that won't be properly tracked.

i.e. The declarative schema changer needs to know what objects this expression refers to, otherwise drops won't work correctly
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_function_param_default.go` line 24 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-NueeMDx-pLKD7ADfbfb:-NueeMDx-pLKD7ADfbfc:b7r8wd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/813b02479cf0aae907867630cf0df3f2463f79e3/pkg/sql/schemachanger/scplan/internal/opgen/opgen_function_param_default.go#L24)):*
> ```Go
> 			to(scpb.Status_PUBLIC,
> 				// TODO(chengxiong): add operations when default value is supported.
> 				emit(func(this *scpb.FunctionParamDefaultExpression) *scop.NotImplementedForPublicObjects {
> ```

Even if we embed it we need a similar concept to track dependencies inside from the default expression.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @fqazi, @mgartner, and @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 681 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-NuecjJr3fwexDQtyLj3:-Nuef_Bg8hmKB9Ee2O9d:b-6mmux4) ([raw file](https://github.com/cockroachdb/cockroach/blob/220f1a84eef868ffa23427ced83aea50ac917cd7/pkg/sql/schemachanger/scbuild/builder_state.go#L681)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

When did we run into this? This shouldn't be a thing :grimacing:
</blockquote></details>

Good catch. This was a leftover from the version where I used `WrapExpression` on the DEFAULT expression, and `tableID` argument was the function ID (because on the interface it is named `parentID`). Removed.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 672 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nued7vuCQt5U-S6a2bD:-Nuf-Sl90Cl53EU7JAVs:bgf6vui) ([raw file](https://github.com/cockroachdb/cockroach/blob/220f1a84eef868ffa23427ced83aea50ac917cd7/pkg/sql/schemachanger/scpb/elements.proto#L672)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This won't work correctly, we can't track dependencies, we need to at least embed an expression directly in parameter. Otherwise, we can have references in the default expression that won't be properly tracked.

i.e. The declarative schema changer needs to know what objects this expression refers to, otherwise drops won't work correctly
</blockquote></details>

Hm, I can't think of an example where things are broken. Right now, in the optbuilder we include dependencies from the DEFAULT expressions into the same set of dependencies from the function body, so all UDTs, UDFs, and sequences referenced from the DEFAULT expressions end up being included into the function body dependencies. Is this insufficient?


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121082)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @rafiss, and @yuzefovich)

___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 672 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/121082#-Nued7vuCQt5U-S6a2bD:-Nuf0p-P4_r7w9LWFPC3:b-tu6roo) ([raw file](https://github.com/cockroachdb/cockroach/blob/220f1a84eef868ffa23427ced83aea50ac917cd7/pkg/sql/schemachanger/scpb/elements.proto#L672)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I can't think of an example where things are broken. Right now, in the optbuilder we include dependencies from the DEFAULT expressions into the same set of dependencies from the function body, so all UDTs, UDFs, and sequences referenced from the DEFAULT expressions end up being included into the function body dependencies. Is this insufficient?
</blockquote></details>

Oh I didn't realize that the existing tracking covered that. Yeah, it's sufficient.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for the reviews everyone!

bors r+

yuzefovich commented:
blathers backport release-24.1
